Determine display DPI from scaling factor

vadz opened the pull request:
See [#18855](https://trac.wxwidgets.org/ticket/18855): with these changes, the returned value of all of both `GetDPIScaleFactor()` and `GetContentScaleFactor()` is 2 when using high DPI screen. @paulcor do we care about always returning 1 here for GTK 3 < 3.10?

Note that this almost surely does *not* work under Mac, as `wxDisplayImpl::GetScaleFactor()` is not implemented there. @csomor Could you please add an implementation for it? It should be pretty simple, but I don't want to do it blindly. TIA!

vadz committed:
Add wxDisplay::GetStdPPIValue() and GetStdPPI()

Provide public way to access the default platform DPI, which was
previously a constant private to src/common/wincmn.cpp.

vadz committed:
Return "logical" PPI from wxDisplay::GetPPI() in wxGTK3

Just scale the standard PPI by the scaling factor instead of trying to
compute the actual physical PPI, as the latter is not useful and is
incompatible with the value used by all the other applications.

vadz committed:
Add wxDisplay::GetScaleFactor()

This is conceptually the same as the ratio of the current DPI to the
standard one, but can be implemented more directly for wxGTK3 and wxOSX
(although the latter doesn't implement it yet).

vadz committed:
Use wxDisplay::GetScaleFactor() in wxWindow::GetDPIScaleFactor()

This can be more direct and efficient than computing the ratio of DPIs
in wxWindow.

csomor commented:
> @csomor Could you please add an implementation for it? It should be pretty simple, but I don't want to do it blindly. TIA!

ok, so while [#18855](https://trac.wxwidgets.org/ticket/18855) was working correctly on macOS, you prefer to go via wxDisplay on macOS as well ? I'll implement the wxDisplay method, but I'd still prefer to avoid the overhead for the simple case. Could we make `GetDPIScaleFactor` virtual ?

csomor committed:
macOS Implementation for wxDisplay::GetScaleFactor

csomor committed:
add a GetFromWindow implementation to wxDisplayFactory

on displays with different resolutions we must make sure we pick one that matches the frame’s contentScale

csomor committed:
macOS clientRect implementation for non-primary displays

MaartenBent commented on interface/wx/display.h:
```diff
@@ -143,6 +143,46 @@ class wxDisplay
+
+        Currently the standard PPI is the same in both horizontal and vertical
+        directions on all platforms and its value is 96 everywhere except under
+        Mac, where it is 72.
```
Maybe it is better to use macOS or wxOSX here?

csomor replied:
yes, since it is both macOS and iOS, perhaps wxOSX would be a good alternative

vadz replied:
Does wxOSX include wxiOS? Maybe we should just say "Apple devices" :-/

I really don't know what's the best way to refer to macOS and iOS collectively, please let me know what do you prefer.

catalinr replied:
It could just be "macOS and iOS". It is clear and they are not that many.

csomor replied:
I think Apple devices is a good idea, watchOS, tvOS would be included as well

MaartenBent commented on interface/wx/display.h:
```diff
@@ -143,6 +143,46 @@ class wxDisplay
+        Returns default display resolution for the current platform as wxSize.
+
+        This function is equivalent to constructing wxSize object with both
+        components set to GetStdPPIValue()/
```
`/` → `.`

vadz commented:
> ok, so while [#18855](https://trac.wxwidgets.org/ticket/18855) was working correctly on macOS, you prefer to go via wxDisplay on macOS as well ?

This seems to make sense to me, but I don't know how is this implemented there. Under MSW it comes from the display and while we do have `gtk_widget_get_scale_factor()` in GTK, its implementation ends up using `gdk_screen_get_monitor_scale_factor()` anyhow, so I thought we could just as well go directly there. Are you sure that Mac doesn't do the same thing internally?

Also, I think we need to cache these values (see the recent post by AliKet to wx-dev), and it seems to make more sense to cache them at `wxDisplay` level, rather than for each `wxWindow`. Of course, this also assumes that finding the display from the window is fast, and I'm not sure about that...

Finally, having the scale factor available in `wxDisplay` may be useful on its own, as sometimes you need it _before_ creating the window.

> I'll implement the wxDisplay method, but I'd still prefer to avoid the overhead for the simple case. Could we make `GetDPIScaleFactor` virtual ?

If we need a different implementation for Mac, yes, sure.

csomor commented:
> This seems to make sense to me, but I don't know how is this implemented there. Under MSW it comes from the display and while we do have `gtk_widget_get_scale_factor()` in GTK, its implementation ends up using `gdk_screen_get_monitor_scale_factor()` anyhow, so I thought we could just as well go directly there. Are you sure that Mac doesn't do the same thing internally?

I have my doubts, contentScale is on the NSWindow level, I guess that one gets updated with every NSWindow frame change

> Also, I think we need to cache these values (see the recent post by AliKet to wx-dev), and it seems to make more sense to cache them at `wxDisplay` level, rather than for each `wxWindow`. Of course, this also assumes that finding the display from the window is fast, and I'm not sure about that...

me neither, at least compared to just use contentScale ... but I've adapted wxDisplay's `FromWindow` to make sure the factor matches

> Finally, having the scale factor available in `wxDisplay` may be useful on its own, as sometimes you need it _before_ creating the window.

yes, definitely, I absolutely agree about having the implementation (done btw) is important

> > I'll implement the wxDisplay method, but I'd still prefer to avoid the overhead for the simple case. Could we make `GetDPIScaleFactor` virtual ?
> 
> If we need a different implementation for Mac, yes, sure.

at least I'd prefer to have the shortcut there, I'll do that then ...

csomor committed:
making GetDPIScaleFactor virtual, adding macOS direct impl

vadz commented on src/osx/core/display.cpp:
```diff
@@ -196,31 +228,44 @@ int wxDisplayFactoryMacOSX::GetFromPoint(const wxPoint& p)
+    for ( int i = 0 ; i < theCount; ++i )
+    {
+        // find a screen intersecting having the same contentScale as the window itself
+        wxCFRef<CGDisplayModeRef> mode = CGDisplayCopyDisplayMode(theIDs[i]);
+        size_t width = CGDisplayModeGetWidth(mode);
+        size_t pixelsw = CGDisplayModeGetPixelWidth(mode);
```
It would be nice to reuse the same code in `wxDisplay::GetScaleFactor()` added in the previous commit and here, e.g. by adding a common helper function.

csomor replied:
yes, I'll do that

vadz commented on src/osx/window_osx.cpp:
```diff
@@ -2629,16 +2629,23 @@ wxSize wxWindowMac::OSXMakeDPIFromScaleFactor(double scaleFactor)
     if ( wxNonOwnedWindow* tlw = MacGetTopLevelWindow() )
         scaleFactor = tlw->GetContentScaleFactor();
     else
         scaleFactor = wxOSXGetMainScreenContentScaleFactor();
```
Should we ask the window display for its scale factor here?

Of course, normally we should always have the containing TLW anyhow, so maybe it's not worth it...

csomor replied:
we should ask the TLW as there we have the 'true' value, the way it is handled by the OS. If there isn't a tlw, there isn't a absolute position either, so I think using the main screen's scale is as good or bad as any other. Or did I not understand you correctly ?

vadz replied:
I'm not sure, could we have a position but no TLW because it wasn't shown (at this position) yet?

vadz replied:
Actually, why don't we just do the same thing (or just call) `GetContentScaleFactor()` here? I don't understand why is there the difference between the two, as the comment says, they should return the same thing?

csomor replied:
yes, they should return the same thing, but I haven't built in a fallback in case there's no tlw, so let's have both go through the same route and I'll add that condition to the code

csomor committed:
using common method for determining scale from displayID

vadz committed:
Implement GetDPIScaleFactor() more efficiently in wxGTK

Don't pass via wxDisplay, but use gtk_widget_get_scale_factor() (almost)
directly instead.

vadz committed:
Implement GetDPIScaleFactor() more directly for wxMSW too

Use the existing GetDPI() implementation instead of forwarding to
wxDisplay.

vadz committed:
Simplify PPI and scale factor handling in wxDisplay

Don't try computing the PPI ourselves from the physical size and the
number of pixels, this doesn't work and nobody else does it like this.

Just assume that we're using standard PPI by default and use
toolkit-specific functions for the platforms with support for high DPI.

vadz committed:
Remove unused wxDisplayImpl::GetSizeMM()

This is not needed any longer after the changes of the last commit.

Note that the (still existent) public wxGetDisplaySizeMM() didn't use
this function, but used PPI instead.

vadz committed:
Show scale factor in the display sample too

Show it for consistency with all the other wxDisplay accessors and to
make it simpler to test its return value.

vadz commented:
I've done the same changes for wxGTK as you did for wxOSX and removed the now unused code dealing with size in mm and computing PPI from it. AFAICS, this could be merged now, although I still don't know where/how to cache the DPI/scale factor values, as we're definitely going to need it more and more and so we really should cache them.

Please let me know if you plan to make any more changes here, otherwise I'll rebase it to get rid of fixups and merge. TIA!

csomor committed:
use common contentScaleFactor also for DPI on macOS

and give a reasonable default for missing tlw

csomor commented:
I've added the change regarding the dpi/contentScale, I'm done now

vadz commented:
Thanks! I've rebased to squash fix ups, so this should be ready to go. I'll wait for a couple of days for any comments, as this is a rather major change (let me specifically ping @MaartenBent, @paulcor, @stbrowne but comments from all the others are welcome too, of course).

MaartenBent reviewed (approved):
MSW changes look good to me.
 I can't really comment on gtk/osx, except for mentioning that Travis fails on a missing `wx/nonownedwnd.h` include.

vadz committed:
Make a minor optimization in Mac wxDisplay::GetFromWindow()

Call GetContentScaleFactor() only once, which doesn't cost anything and
could be minutely faster on systems with multiple displays.

vslavik commented:
This change is causing crashes on macOS (e.g.) when creating `wxTextCtrl` here:

```
#0	0x0000000100c908d5 in wxWindow::GetContentScaleFactor() const at src/osx/window_osx.cpp:1205
#1	0x0000000100c95f09 in wxWindow::GetDPIScaleFactor() const at src/osx/window_osx.cpp:2639
#2	0x0000000100c95ed9 in wxWindow::GetDPI() const at src/osx/window_osx.cpp:2633
#3	0x0000000100fc1ea9 in wxWindowBase::GetFont() const at src/common/wincmn.cpp:1713
#4	0x0000000100dfc27f in wxNSTextViewControl::SetStringValue(wxString const&) at src/osx/cocoa/textctrl.mm:809
#5	0x0000000100e0059e in wxWidgetImpl::CreateTextControl(wxTextCtrl*, wxWindow*, int, wxString const&, wxPoint const&, wxSize const&, long, long) at src/osx/cocoa/textctrl.mm:1595
#6	0x0000000100d38255 in wxTextCtrl::Create(wxWindow*, int, wxString const&, wxPoint const&, wxSize const&, long, wxValidator const&, wxString const&) at src/osx/textctrl_osx.cpp:102
#7	0x0000000100222e81 in wxTextCtrl::wxTextCtrl(wxWindow*, int, wxString const&, wxPoint const&, wxSize const&, long, wxValidator const&, wxString const&) at include/wx/osx/textctrl.h:46
```

I'm not sure what an appropriate fix would be. The sequence of events is:

1. `wxTextCtrl::Create()` calls creation of the native peer object via `wxWidgetImpl::CreateTextControl()`; note that at this point `m_peer` is still unset
2. During that, `wxNSTextViewControl::SetStringValue()` gets called and it in turn calls `m_wxPeer->GetFont()` …
3. … which now does some PPI adjustment, which eventually calls `wxWindowMac::GetContentScaleFactor()` which requires the peer object, which is still unset, see 1.

How to address it is unclear to me. I would be vary of doing it for `wxTextCtrl` only as such hidden dependencies may be (and probably are) present elsewhere in the code.

Frustratingly, this crash is an issue only because the font is needed for its encoding, which is only needed in long-deprecated `wxUSE_UNICODE=0` builds (which are broken since Jul 12 due to b2d4a9d...).

csomor commented:
Thanks, I'll look at that. What I didn't understand was the sequence. Why do you need `wxUSE_UNICODE=0`?

csomor commented:
GetDPI is asking for the toplevel window and its peer is used, so I assume in your case the framewindow is not yet created ?

The PPI adjustments are done only on msw, so that call in wxWindowBase::GetFont is actually unnecessary 

I'd suggest making GetContentScaleFactor non-peer proof:

```
double wxWindowMac::GetContentScaleFactor() const 
{
    if ( GetPeer() )
        return GetPeer()->GetContentScaleFactor();
    else
        return wxOSXGetMainScreenContentScaleFactor();
}
```

what do you think ?

vslavik commented:
> Why do you need wxUSE_UNICODE=0?

I don't (\*cringes in horror\*). The code crashes in Unicode build too, because it calls `GetFont` regardless, but it does it *pointlessly* in Unicode builds: the encoding in `wxCFStringRef(st , m_wxPeer->GetFont().GetEncoding())` is unused and `wxCFStringRef(st)` would be sufficient (and wouldn't call `GetFont()` and so crash).

> GetDPI is asking for the toplevel window and its peer is used, so I assume in your case the framewindow is not yet created ?

Err, no, `wxWindow::GetDPI()` calls everything on the same window, including `wxWindow::GetContentScaleFactor()` (and my code creates everything normally).

> The PPI adjustments are done only on msw, so that call in wxWindowBase::GetFont is actually unnecessary
…
> what do you think ?

I would feel uneasy about special-casing early construction like this… If the adjustment isn't done on other platforms and `WXAdjustToPPI` is a noop (I see it now), why do the busywork in the first place? I'd say this does call for an `#ifdef` so that `GetPPI()` doesn't keep getting called only to throw out the return value.

P.S., @vadz has certainly a much better understanding of what users' needs are these days and if ANSI build still happens, but personally I would be inclined to say that it is time now — 7 years after 3.0's release where `wxUSE_UNICODE` was deprecated — to remove it altogether, which would also happen to fix this issue. This PR is not the place for that discussion, but FWIW, I'd be happy to do the work.

vadz commented:
The trouble with using the main screen scale factor as fallback in `wxWindow` is that it will return a wrong value for a window which is about to appear on another screen. This is not as fatal as crashing, of course, but will be even more difficult to debug...

Anyhow, fixing this particular crash under Mac is simple: we should clearly skip claling `GetDPI()` there in any case, because it's perfectly useless when `WXAdjustToPPI()` doesn't do anything there. IOW we should have `WXAdjustToOwnPPI(void)` and keep the existing `WXAdjustToPPI()` only in wxMSW and other ports where it's needed (i.e. not wxGTK3/wxOSX). We'd still have a similar problem in wxMSW, but, again, it would be "just" wrong there, and won't crash. @MaartenBent do you agree with the proposed change?

Re `wxUSE_UNICODE=0`: officially we still support it up to 3.2.0 (I'll rip it off afterwards), so it would be great if this could be fixed until then. @vslavik could you please open a Trac ticket with 3.1.5 milestone if you can't do it?

P.S. Vaclav's reply appeared while I was writing this and, unfortunately, I still know of at least one project which has never upgraded to build with `wxUSE_UNICODE=1`. And there are questions about it on the mailing list from time to time, so people still use it, although don't ask me why.

vslavik commented:
> IOW we should have WXAdjustToOwnPPI(void) and keep the existing WXAdjustToPPI() only in wxMSW

Done, although it's a bit more complicated — see #2036.

Thanks for the discussion, it helped my understanding a lot.
